### PR TITLE
[Issue #128] software: add start/stop buttons handling

### PR DESCRIPTION
**Pull request by csommelet**
Implement issue #128

Still work in progress:
- not sure this is the right design: it starts to be hard to follow the current state of the device and the various interactions between the components (display, alarms, etc.)
- for now, we wait the end of the current cycle before stopping the breathing. Should we stop in the middle of the cycle?
- reset actions to do on stop are still to be implemented

### CODE REVIEW

File: src/software/firmware/srcs/respirator.cpp [hunk 1]
```diff
@@ -217,4 +218,50 @@ void loop() {
...
+    // Just wait one compute period.
+    uint32_t lastpControllerComputeDate = 0uL;
+    while (true) {
+        uint32_t currentDate = millis();
+        if ((currentDate - lastpControllerComputeDate) >= PCONTROLLER_COMPUTE_PERIOD) {
+            break;
+        }
+        lastpControllerComputeDate = currentDate;
+        IWatchdog.reload();
+    }
```
Comment (ThomasL01) [inline src/software/firmware/srcs/respirator.cpp:240]: Is there a possiblility that (currentDate - lastpControllerComputeDate) >= PCONTROLLER_COMPUTE_PERIOD) stays false. Then updating lastpControllerComputeDate with currentDate in the next line will create a non return condition continuously kipping stuck in the while(true). 
IWatchdog keeps reloading and no button can be triggered in that case the device needs to be power off/on to force reboot, right?

Comment (csommelet) [reply src/software/firmware/srcs/respirator.cpp:240]: I kept the logic used in the main respiratory cycle loop. I dont know what is the risk that millis() returns incoherent data.

Comment (ThomasL01) [reply src/software/firmware/srcs/respirator.cpp:240]: Yes I understand the logic. millis() is returning the number of milliseconds passed since the program started which is fine and can last for 50 days.
I am afraid that the loop will loop forever with the statement you use after lastpControllerComputeDate = currentDate; because it won't reach (currentDate - lastpControllerComputeDate) >= PCONTROLLER_COMPUTE_PERIOD (//10ms here)
If it is the behaviour you want, than it is fine to me.

**Comment by jabby** [comment]
>     * for now, we wait the end of the current cycle before stopping the breathing. Should we stop in the middle of the cycle?

I asked @Mefl 
We should not stop in the middle of the cycle.

File: src/software/firmware/srcs/respirator.cpp (hunk 1 above)
Comment (jabby) [reply src/software/firmware/srcs/respirator.cpp]: @ThomasL01 you're right. The loop looped forever. I change the behaviour and merged both methods. 
We want to avoid duplication of some part like watchdog reload, keyboardLoop and batteryLoop.

**Comment by jabby** [comment]
@csommelet I update you're PR with latest master. Feel free to review my changes. 
By the way, thanks for your work. It help me a lot.

File: src/software/firmware/includes/activation.h [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+#pragma once
+
+// INCLUDES ===================================================================
+
+// External libraries
+#include <Arduino.h>
+
+// CLASS ======================================================================
+
+/// Controls breathing activation on/off state
```
Comment (dsferruzza) [inline src/software/firmware/includes/activation.h:17]:
```suggestion
/// Controls breathing activation ON/OFF state
```

File: src/software/firmware/includes/activation.h [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    void onStartButton();
+
+    /**
+     * Callback to call each time the stop button is pushed
+     */
+    void onStopButton();
+
+ private:
+    enum State {
+        STOPPED = 0,            // Breathing is off
```
Comment (dsferruzza) [inline src/software/firmware/includes/activation.h:39]:
```suggestion
        STOPPED = 0,            // Breathing is OFF
```

File: src/software/firmware/includes/activation.h [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        RUNNING,                // Breathing is on
```
Comment (dsferruzza) [inline src/software/firmware/includes/activation.h:40]:
```suggestion
        RUNNING,                // Breathing is ON
```

File: src/software/firmware/includes/activation.h [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        RUNNING_READY_TO_STOP,  // Breathing is on, waiting for a second push
```
Comment (dsferruzza) [inline src/software/firmware/includes/activation.h:41]:
```suggestion
        RUNNING_READY_TO_STOP,  // Breathing is ON, waiting for a second push
```

File: src/software/firmware/srcs/activation.cpp [hunk 6]
```diff
@@ -0,0 +1,50 @@
+/******************************************************************************
+ * @author Makers For Life
+ * @copyright Copyright (c) 2020 Makers For Life
+ * @file activation.h
```
Comment (dsferruzza) [inline src/software/firmware/srcs/activation.cpp:4]:
```suggestion
 * @file activation.cpp
```

File: src/software/firmware/srcs/activation.cpp [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+ * @brief Breathing activation related functions
+ *****************************************************************************/
+
+// INCLUDES ===================================================================
+
+// Associated header
+#include "../includes/activation.h"
+
+// External libraries
```
Comment (dsferruzza) [inline src/software/firmware/srcs/activation.cpp:13]:
```suggestion
```

File: src/software/firmware/srcs/activation.cpp [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+// Internal
+#include "../includes/buzzer.h"
+
+const uint32_t SECOND_STOP_MAX_DELAY_MS = 5000;
```
Comment (dsferruzza) [inline src/software/firmware/srcs/activation.cpp:18]:
```suggestion
/// Maximum delay in seconds between two pushes on the stop button
/// to actually stop the machine
const uint32_t SECOND_STOP_MAX_DELAY_MS = 5000;
```

File: src/software/firmware/srcs/screen.cpp [hunk 9]
```diff
@@ -165,3 +163,10 @@ void displayAlarmInformation(uint8_t p_alarmCodes[], uint8_t p_nbTriggeredAlarms
     screen.setCursor(0, 2);
     screen.print(buf);
 }
+
+void displayMachineStopped(void) {
+    screen.setCursor(0, 3);
+    char message[SCREEN_LINE_LENGTH + 1];
+    (void)snprintf(message, SCREEN_LINE_LENGTH + 1, "%-20s", "MakAir stopped.");
+    screen.print(message);
```
[the remark this answers is not in the document]
Comment (dsferruzza) [reply src/software/firmware/srcs/screen.cpp:171]: Also display the version

File: src/software/firmware/srcs/activation.cpp [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    switch (m_state) {
+    case STOPPED:
+        break;
+
+    case RUNNING_READY_TO_STOP:
+        if (millis() - m_timeOfLastStopPushed < SECOND_STOP_MAX_DELAY_MS) {
+            m_state = STOPPED;
+            break;
+        }
+        // else fallback to default ON state handling
```
Comment (dsferruzza) [inline src/software/firmware/srcs/activation.cpp:40]: Yes please

File: src/software/firmware/srcs/activation.cpp [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+    case RUNNING:
+        m_timeOfLastStopPushed = millis();
+        m_state = RUNNING_READY_TO_STOP;
+
+        // TODO check what to do in case of running alarm
```
Comment (dsferruzza) [inline src/software/firmware/srcs/activation.cpp:46]: Not sure what we should do here? Nothing?

File: src/software/firmware/srcs/activation.cpp [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+// INITIALISATION =============================================================
+
+ActivationController activationController;
+
+// FUNCTIONS ==================================================================
+
+ActivationController::ActivationController() : m_state(STOPPED), m_timeOfLastStopPushed(0) {}
+
+void ActivationController::onStartButton() { m_state = RUNNING; }
```
Comment (dsferruzza) [inline src/software/firmware/srcs/activation.cpp:28]: If previous state was `STOPPED`, maybe we should do a `Buzzer_Boot_Start()`?

File: src/software/firmware/srcs/blower.cpp [hunk 13]
```diff
@@ -50,5 +50,4 @@ int Blower::getSpeed() { return m_speed; }
 
 void Blower::stop() {
     actuator->setCaptureCompare(timerChannel, BlowerSpeed2MicroSeconds(0), MICROSEC_COMPARE_FORMAT);
-    m_speed = 0;
```
[the remark this answers is not in the document]
Comment (jabby) [reply src/software/firmware/srcs/blower.cpp:53]: We don't set m_speed to 0 because we want to keep the latest ajustment of blower speed. 
The real stop is `actuator->setCaptureCompare(timerChannel, BlowerSpeed2MicroSeconds(0), MICROSEC_COMPARE_FORMAT);`

[the remark this answers is not in the document]
Comment (dsferruzza) [reply src/software/firmware/srcs/blower.cpp:53]: 👌

File: src/software/firmware/srcs/respirator.cpp [hunk 14]
```diff
@@ -125,17 +116,25 @@ void setup() {
...
 
-    // escBlower start
-    blower.runSpeed(150);
+    lastpControllerComputeDate = millis();
+
+    /* Catch potential Watchdog reset */
+    if (IWatchdog.isReset(true)) {
+        // TODO holdExhale
+        // Display something ?
+        // Alarm code ?
```
[the remark this answers is not in the document]
Comment (jabby) [reply src/software/firmware/srcs/respirator.cpp:153]: I just move this code as I think it's better to finish initialization before launch watchdog alarm.
I plan to create another PR with missing functionnality.
I need to validate behaviour with @cmdm

File: src/software/firmware/srcs/screen.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (jabby) [reply src/software/firmware/srcs/screen.cpp:171]: First and second line are used for settings.
Third one is used for alarm.
I have only 20 characters so I suggest `Press start to begin`.
It suggest that the device is stopped and explain how to do it.

File: src/software/firmware/srcs/screen.cpp [hunk 15]
```diff
@@ -165,3 +163,10 @@ void displayAlarmInformation(uint8_t p_alarmCodes[], uint8_t p_nbTriggeredAlarms
     screen.setCursor(0, 2);
     screen.print(buf);
 }
+
+void displayMachineStopped(void) {
+    screen.setCursor(0, 3);
+    char message[SCREEN_LINE_LENGTH + 1];
+    (void)snprintf(message, SCREEN_LINE_LENGTH + 1, "%-20s", "MakAir stopped.");
```
[the remark this answers is not in the document]
Comment (jabby) [reply src/software/firmware/srcs/screen.cpp:170]: `Press start to begin` seems to be a good message. I just removed runtime composition.

File: src/software/firmware/srcs/respirator.cpp [hunk 16]
```diff
@@ -125,17 +116,25 @@ void setup() {
...
     digitalWrite(PIN_LED_RED, LOW);
     digitalWrite(PIN_LED_YELLOW, LOW);
-
     waitForInMs(4000);
 
-    // escBlower start
-    blower.runSpeed(150);
+    lastpControllerComputeDate = millis();
+
+    /* Catch potential Watchdog reset */
```
[the remark this answers is not in the document]
Comment (jabby) [reply src/software/firmware/srcs/respirator.cpp:149]: Done

File: src/software/firmware/includes/activation.h (hunk 2 above)
Comment (jabby) [reply src/software/firmware/includes/activation.h:17]: Done

File: src/software/firmware/includes/activation.h (hunk 3 above)
Comment (jabby) [reply src/software/firmware/includes/activation.h:39]: Done

File: src/software/firmware/includes/activation.h (hunk 4 above)
Comment (jabby) [reply src/software/firmware/includes/activation.h:40]: Done.

File: src/software/firmware/includes/activation.h (hunk 5 above)
Comment (jabby) [reply src/software/firmware/includes/activation.h:41]: Done

File: src/software/firmware/srcs/activation.cpp (hunk 6 above)
Comment (jabby) [reply src/software/firmware/srcs/activation.cpp:4]: Done.

File: src/software/firmware/srcs/activation.cpp (hunk 7 above)
Comment (jabby) [reply src/software/firmware/srcs/activation.cpp:13]: Done.

File: src/software/firmware/includes/screen.h [hunk 17]
```diff
@@ -71,5 +71,10 @@ void displayCurrentInformation(uint16_t peakPressure, uint16_t plateauPressure,
  */
 void displayAlarmInformation(uint8_t p_alarmCodes[], uint8_t p_nbTriggeredAlarms);
 
+/**
+ * Display machine stopped.
```
[the remark this answers is not in the document]
Comment (jabby) [reply src/software/firmware/includes/screen.h:75]: Done

[the remark this answers is not in the document]
Comment (jabby) [reply src/software/firmware/includes/screen.h:75]: Done

File: src/software/firmware/srcs/activation.cpp (hunk 11 above)
Comment (jabby) [reply src/software/firmware/srcs/activation.cpp:46]: I think we should do nothing.
I removed the comment.

File: src/software/firmware/srcs/activation.cpp (hunk 10 above)
Comment (jabby) [reply src/software/firmware/srcs/activation.cpp:40]: Update comment as it's not an obvious behavior.

File: src/software/firmware/srcs/activation.cpp (hunk 12 above)
Comment (jabby) [reply src/software/firmware/srcs/activation.cpp:27]: Actually there is no need.
